dali/pipeline/executor/executor2/exec_graph.h

```diff
@@ -0,0 +1,362 @@
+   */
+  std::shared_ptr<tasking::Semaphore> concurrency;
+
+  /** Limits the number of output buffers for the operator.
+   *
+   * If the graph is scheduled multiple times ahead, the operator would produce multiple results,
+   * potentially consuming a lot of memory. The output_queue_limit limits the number of pending
+   * outputs. Once all consumers are done with previous iterations, the semaphore is released
+   * and the task may proceed with the next iteration.
+   */
+  std::shared_ptr<tasking::Semaphore> output_queue_limit;
+
+  /** The instance of the operator (or null for output node) */
+  std::unique_ptr<OperatorBase> op;
+
+  /** The task from the previous iteration - kept in order to maintain execution order */
+  tasking::SharedTask prev;
+
+  /** The task from the current iteration */
+  tasking::SharedTask main_task;
+
+  /** The task that releases the output_queue_limit semaphore.
+   *
+   * IMPORTANT: Release_outputs is NOT required for correctness. It's sole function is limiting
+   *            the number of pending outputs from the operator.
+   *
+   * This is an auxiliary task which is scheduled after all direct successors of main_task.
+   *
+   * op A --- (output 0) ----- op B --------- ...
+   *    \                         /
+   *     ------- (output 1) -----(- op C --- ...
+   *                              \   |
+   *                               \  |
+   *                                \  \
+   *                                 \_release outptus (A)
+   */
+  tasking::SharedTask release_outputs;
+
+  /** Data-independent execution environment (thread pool, stream, etc). */
+  ExecEnv env = {};
+
+  /** Places a workspace in a workspace cache. */
+  void PutWorkspace(CachedWorkspace ws);
+
+  /** Obtains a worskpace from a workspace cache or, if not found, creates a new one. */
+  CachedWorkspace GetWorkspace(std::shared_ptr<IterationData> iter_data,
+                               WorkspaceParams params);
+
+  /** The instance name of the operator. */
+  const std::string instance_name;
+
+  /** The backend on which the operator runs. */
+  OpType backend = OpType::CPU;
+
+  /** Whether the node is the very output of the pipeline. There's only one such node. */
+  bool is_pipeline_output = false;
+
+  /** Whether the operator in the node is a batch size provider. */
+  bool is_batch_size_provider = false;
+
+  /** Visit marker for graph algorithms. */
+  mutable bool visited = false;
+
+ private:
+  /** Creates a workspace at the pipeline's output.
+   *
+   * Output node's inputs are converted to Pipeline's outputs. The output workspace doesn't have
+   * any inputs.
+   */
+  CachedWorkspace CreateOutputWorkspace();
+  /** Creates a workspace suitable for running the operator `op`. */
+  CachedWorkspace CreateOpWorkspace();
+
+  WorkspaceCache workspace_cache_;
+
+  /** Moves to a new iteration. */
+  void NextIter() {
+    prev = std::move(main_task);
+    release_outputs.reset();
```

Comment:
I've yet to read this file properly, but the number of public members here worry me a bit.

This makes reading the methods difficult - as the members' names do not reflect the fact they are members. And, of course, the problem that those members may be modified externally in uncontrolled manner.

Reply:
Well, they _are_ modified externally, e.g. with the graph analyzer. Adding access functions that do nothing isn't adding to the program's readability. About naming - well, it's kind of _because_ the members are publicly available (much like in a struct). I can reconsider access qualification of some of those members - perhaps not everything that's now public needs to be.